### PR TITLE
Update rule selection to respect preview mode

### PR DESCRIPTION
**Pull request by zanieb**
## Summary

Extends work in #7046 (some relevant discussion there)

Changes:

- All nursery rules are now referred to as preview rules
- Documentation for the nursery is updated to describe preview
- Adds a "PREVIEW" selector for preview rules
    - This is primarily to allow `--preview --ignore PREVIEW --extend-select FOO001,BAR200`
- Using `--preview` enables preview rules that match selectors

Notable decisions:

- Preview rules are not selectable by their rule code without enabling preview
- Retains the "NURSERY" selector for backwards compatibility
- Nursery rules are selectable by their rule code for backwards compatiblity

Additional work:

- Selection of preview rules without the "--preview" flag should display a warning
- Use of deprecated nursery selection behavior should display a warning
- Nursery selection should be removed after some time

## Test Plan

Manual confirmation (i.e. we don't have an preview rules yet just nursery rules so I added a preview rule for manual testing)

New unit tests

### CODE REVIEW

File: crates/ruff_benchmark/benches/linter.rs
Comment (zanieb) [inline crates/ruff_benchmark/benches/linter.rs:82]: Presumably the benchmark regressions are due to including our nursery/preview rules now

Comment (zanieb) [reply crates/ruff_benchmark/benches/linter.rs]: Yep https://github.com/astral-sh/ruff/pull/7195/commits/9522dfbc0119727b3c8ec8a7b0991159c6229c8c — I'll do this in a separate PR to avoid blocking this one

Comment (zanieb) [reply crates/ruff_benchmark/benches/linter.rs]: #7208

File: crates/ruff_dev/src/generate_docs.rs
Comment (charliermarsh) [inline crates/ruff_dev/src/generate_docs.rs:48]: Is the "explicitly selecting" part still true? I thought this applied to the existing nursery rules for backwards compatibility, but not for preview rules in general?

File: crates/ruff/src/codes.rs
Comment (charliermarsh) [inline crates/ruff/src/codes.rs:58]: Can we mark as `#[deprecated(...)]` to avoid accidental usages in the future?

File: crates/ruff/src/rule_selector.rs
Comment (charliermarsh) [inline crates/ruff/src/rule_selector.rs:106]: I bet this could be encoded in the type system somehow, this feels awkward (I know that I wrote it lol). Not asking for a change here since it will likely involve more extensive changes in the macro, etc., just expressing that it feels strange.

**Review by charliermarsh — APPROVED** [review]

**Review by charliermarsh — COMMENTED** [review]
Just confirming (even though it's mentioned in the test plan) that no rules are actually moved to `--preview` here -- it's only enabling the functionality.

File: crates/ruff_dev/src/generate_docs.rs
Comment (zanieb) [reply crates/ruff_dev/src/generate_docs.rs:48]: 👍 yep thanks will fix

File: ruff.schema.json
Comment (zanieb) [reply ruff.schema.json:1888]: Yep #7210 will show a warning

**Comment by zanieb** [comment]
@charliermarsh well all of the nursery rules will be enabled with `--preview` but yes there are no rules that are preview-only without being in the nursery.

File: crates/ruff_macros/src/map_codes.rs
Comment (MichaReiser) [inline crates/ruff_macros/src/map_codes.rs:133]: We could move all of this logic out of the macro if we instead implement `impl From<#linter> for RuleCodePrefix` because the `RuleCodePrefix::#linter(linter)` code is the only code that requires code-gen

File: crates/ruff_cli/src/commands/rule.rs
Comment (zanieb) [reply crates/ruff_cli/src/commands/rule.rs:22]: The rule is in preview and is enabled or disabled by preview mode :D I want to think a bit more about "preview" vs "preview mode" in general so thanks for pointing this out.

File: crates/ruff_macros/src/map_codes.rs
Comment (zanieb) [reply crates/ruff_macros/src/map_codes.rs:133]: Interesting. I'd like to separate further changes to the macro from this pull request though, if that's okay?

File: crates/ruff/src/rule_selector.rs
Comment (zanieb) [reply crates/ruff/src/rule_selector.rs:108]: I will open a tracking issue once this is merged

**Comment by zanieb** [comment]
I'd like to do a release tomorrow before merging this.

File: crates/ruff/src/rule_selector.rs
Comment (dhruvmanila) [inline crates/ruff/src/rule_selector.rs:338]:
```suggestion
    /// The specificity when selecting via a rule prefix at three-character depth (e.g., `--select PLE120`).
    Prefix3Chars,
```

File: crates/ruff/src/rule_selector.rs
Comment (dhruvmanila) [inline crates/ruff/src/rule_selector.rs:340]:
```suggestion
    /// The specificity when selecting via a rule prefix at four-character depth (e.g., `--select PLE120`).
    Prefix4Chars,
```

File: crates/ruff/src/rule_selector.rs
Comment (dhruvmanila) [inline crates/ruff/src/rule_selector.rs:334]: I got a bit confused with "rule prefix" which made me think of the _prefix_ part of the code i.e., "PLE" in the given example. But I think this is related to the number part, right? Maybe "Suffix1Char"?

**Review by dhruvmanila — APPROVED** [review]

File: crates/ruff_macros/src/map_codes.rs
Comment (MichaReiser) [reply crates/ruff_macros/src/map_codes.rs:133]: Yeah sure. I should have written this more clearly. This is feedback based on last week's discussion on how it probably would be possible to move a lot out of the macros if we abstracted the dynamic parts by traits instead of where we generate the implementations for our types. But this isn't something that concerns this PR. I just noticed it as a good example because the dynamic part is limited to a single line.

**Pull request merged**
